### ISSUE TITLE
Can't use new LND/RTL Channel Backup functionality

### ISSUE THREAD

**Issue by maltokyo** [btcpayserver/btcpayserver-docker#179]
LND and RTL now allow the backups of all channels, and in BTCPay Server it even looks like it works out of the box in RTL (it tells that backups are on and "verified", a false sense of security).

I asked here whether the mnemonic is required to use channel backups and recover from them, and the answer is yes.
https://github.com/lightningnetwork/lnd/issues/3371#issuecomment-518237883

But, as here, we can't use this (without hacks) on BTCPay server - any chances to have backups available by default?
https://github.com/btcpayserver/btcpayserver-docker/issues/167

**Comment by Kukks** [comment]
We have no way to migrate existing deployments in a sane and humane way and LND won't help even though there is access to the private key that is generated from the seed( and is equivalent enough to be a replacement). 
It will require a lot of hacking, hassle and non-fun code in go

**Comment by maltokyo** [comment]
At least then you should disable the “backup” screens in RTL.. super false sense of security.

**Comment by pavlenex** [comment]
That's an issue for RTL repo. I disagree it's "super-false" sense of security, upon deploying LN in BTCPay, we have a gigantic warning about backups.

**Comment by maltokyo** [comment]
OK, thank you for your reply.  For me, I know the risk now, so I am fine.  But if I didn't read all your docs (most people) I would assume that backups were enabled with a recent upgrade, and feel "safe".

**Comment by Roasbeef** [comment]
> We have no way to migrate existing deployments in a sane and humane way and LND won't help even though there is access to the private key that is generated from the seed( and is equivalent enough to be a replacement).

It's not that we won't help. It's that from the _very beginning_ we've informed y'all that users MUST have access to their seed. BTCPay is putting users at risk by not storing, nor presenting the seed to the end user. I've talked with Dorier several times about safe ways to implement this, however afaict, y'all have chosen not to do so. 

A similar case applies to the usage of pruned nodes, which aren't fully supported. 

All of this can easily be resolved with the following steps:
  1. BTCPay starts to store the user's seed externally in it's database, encrypted even, or prompt the user to provide a passphrase for their seed. 
  2. BTCPay asks users to migrate to the new version, and _publicly_ states that on-chain funds (for lnd), in the current format, as well as channels (you can use SCBs) are not able to be recovered in the case of a disaster. 
  3. BTCPay deprecates the old format  

If this project cares about the safety of their users, then once again, I **strongly** recommend that y'all carry out the step outlined above. I will make my self available to answer any questions, and help with communication so we can ensure the the funds of users using BTCPay are safu. 

>  But if I didn't read all your docs (most people) I would assume that backups were enabled with a recent upgrade, and feel "safe".

This is an excellent point! BTCPay hides _critical_ safety information from the end user. Most users using the system today likely aren't aware of it. This should be an _explicit_ part of on boarding. 

> It will require a lot of hacking, hassle and non-fun code in go

There is no Go involved in the steps I outlined above.

**Comment by maltokyo** [comment]
@NicolasDorier , your view? 
I love your project. Don't get me wrong. It’s how I got into lightning! But I do feel the above needs addressing. But I won’t add anymore funds on BTCPayServer for now than I’m prepared to lose if my hard drive crashes.

**Comment by NicolasDorier** [comment]
We spoke several time on it, the problem is that we don't have an easy way to migrate our user without locking them out of their funds when they update, which is not acceptable.

I would either: 
1. Remove the static channel backup feature so people are not under false sense of security.
2. Ask them to backup their wallet file manually (I guess it has the important keys inside)
3. If possible, allowing them to export their xpriv key, and modify our fork of LND to allow using the xpriv instead of mnemonic (I don't know if technically possible @Roasbeef might tell me)

**Comment by maltokyo** [comment]
I guess people would be ok with any manual "opt-in" option here, to switch it on and have everything recoverable.

You will eventually have to think of this in your project anyway right?  All of the users that you are afraid of locking out, will get irate if they lose their funds anyway with hardware failure etc.  Better to at least have the official well-documented option to turn backups on.

**Comment by NicolasDorier** [comment]
@maltokyo that might be a good idea to have an opt-in. I will think about it.

**Comment by NicolasDorier** [comment]
@maltokyo well the difference is that user getting reckt on lightning does not happen everyday, but if I lock people out of their funds during an update I will get 2000 people coming on the chat in panic mode where I can only help them running arcane linux command they likely don't understand.

**Comment by NicolasDorier** [comment]
Feel free to work on it, normally it can work with a new docker fragment.

**Comment by NicolasDorier** [comment]
Ok giving more thought on this the best approach is to modify our fork:

1. Remove the "noseedbackup" setting
2. Assume same behavior as "noseedbackup=1"
3. When a new wallet is created, save the seed mnemonic next to the wallet.db
4. Expose the seed mnemonic file through BTCPay.

**Comment by maltokyo** [comment]
@NicolasDorier - sounds awesome!!!

**Comment by maltokyo** [comment]
@NicolasDorier Would I have to migrate my installation to a brand new installation and start from scratch with channels etc, with your approach above?

**Comment by Kukks** [comment]
Start from scratch in terms of lnd since we have no migration path

**Comment by maltokyo** [comment]
So your current 2000+ (?) users will never have a way to backup unless establishing all channels from scratch? 
Again here some sort of path to backing up (even if it’s a manual pain in the arse) would be optimal. As long as it’s less of a pain than re-establishing all of those channels etc, people will still prefer it I think).

**Comment by woutersamaey** [comment]
What about new users who haven't used LND before? They don't have any funds at risk but want to start out in the best way. Should new users hold off for a little while because of what's discussed here?

**Comment by woutersamaey** [comment]
I have a feeling that the solution lies in making LND multi-user / multi-tenant. Maybe with docker we can run an old LND and a new LND side-by-side and have a migration path? (i.e. move funds from old to new - channels be damned, only after the user consents that is).
Just brainstorming at this point. If it doesn't help, I'll shut up 🤐

**Comment by Roasbeef** [comment]
> We spoke several time on it, the problem is that we don't have an easy way to migrate our user without locking them out of their funds when they update, which is not acceptable.


Please re-read my initial message. There is no lock out, they simply upgrade, BTCPay detects that they're on an older version, then they sweep their funds away. Simple. BTCPay needs to own this mistake and communicate this to all the users, telling them that their funds **are not safe**.

> Remove the static channel backup feature so people are not under false sense of security.

Users that use it do not have a false sense of security. It is BTCPay that is at fault here, for _not giving the user a seed to their Bitcoin wallet_.  

> So your current 2000+ (?) users will never have a way to backup unless establishing all channels from scratch?

That's the current case. Note that I communicated this issue to the BTCPay developers several months ago, and before that, even some time last year. By waiting longer and longer to resolve this issue, they increase the number of affected users. This could have been solved over a year ago, but they chose not to.

**Comment by Roasbeef** [comment]
Also this isn't about users getting rekt or not. It's about Bitcoin
software that doesn't give users any access to their seed at all. Very
early on in the lifetime of the project I explained this. IMO, in 2019, it
is negligent to ship Bitcoin software that doesn't allow users to restore
their funds, and also doesn't surface this information to the user.

If the project is finally fixing this, then I am very happy, as it is long
over due.

**Comment by Roasbeef** [comment]
> Ok giving more thought on this the best approach is to modify our fork:

To achieve this, you don't need to modify the fork. When you _initialize_ `lnd`, you're given the seed right there. As a result, BTCPay can show it directly to the user, or you can store it yourselves.

**Comment by maltokyo** [comment]
Sorry to jump in again. But is the seed needed only for the on-chain wallet? Would this work?

1. Sweep all LND on-chain funds to another wallet to have zero on-chain funds in LND
2. Turn back on the settings to generate the seed when creating a new wallet
3. Use “createaddress” (or is it “newaddress”?) to make a new LND wallet
4. Use this new wallet exclusively going forward.  Would the new wallet play nicely with existing channels, funded by the old wallet?

Sorry if this is stupid. I don’t know how the lnd on-chain wallet interacts with channels, but I do see that lncli allows us to make new wallets.. so just trying to find a workaround.

**Comment by Eskyee** [comment]
@rockstardev brain storming session needed... here  !! I agree we need to provide the best BTCPay Bitcoin & Lightning software out. here, safe n simple, Noob. ready, but Excellent team work guys. n gals xx  its true, BTCPay is growing up very fast... ps Thanks @Roasbeef we need your NRG man!! x

**Comment by NicolasDorier** [comment]
@woutersamaey @maltokyo 

> So your current 2000+ (?) users will never have a way to backup unless establishing all channels from scratch?

> What about new users who haven't used LND before? 

For the existing users, things will continue to work, but they won't have the seed. We can publish a blog post advising them to update at their leasure with proper guidance. (Closing channels and wipeout LND directory)

For new user, they will get the seed.

@Roasbeef 

>  they simply upgrade, BTCPay detects that they're on an older version, then they sweep their funds away. 

And how do they sweep their funds if LND refuse to start in the first place?

**Comment by Roasbeef** [comment]
> And how do they sweep their funds if LND refuse to start in the first place?

Why would it refuse to start? You can detect which version of `lnd` it is (based on the deployment) and conditionally apply the flag. 

> and we don't have resources to spend on this, not on support and not on UX dev to fix this specific problem.

They don't need to do anything on the command line with what I outlined. I understand the project has limited resources, but I would consider the **safety of user funds** to be a _very_ high priority for _any_ Bitcoin project that involves a wallet or keys. 

> Having to do a whole UX that stop LND, sweep funds, change the config, restart LND is an architectural change that alone would take me a month to implement. You telling me that it is my problem and I am at fault don't solve the situation: I don't have time to work on it.

You've known for nearly a year how big an issue this is. If you don't care about the safety of the funds that users put into the project, then I guess you should get some "SFYL" messages ready, or maybe a FAQ to explain why they lost their money, and how it _could've_ been prevented. 

The solution you outlined above IS NOT A SOLUTION. You're putting far too much burden on the user, when we expose the tools in `lnd` to do everything _correctly_. Once again, from the very start, I informed you how _dangerous_ not giving users their _back up seed_ was. If BTCPay fails to act on this, more than a year later, then I think that says a lot about y'all's commitment to y'all's users. 

Yes, it will take some time to do this properly, but isn't it worth it to have the peace of mind that users will be able to restore their funds?

**Comment by cfromknecht** [comment]
> Adding a way in BTCPay to see the seed is easy. Having to do a whole UX that stop LND, sweep funds, change the config, restart LND is an architectural change that alone would take me a month to implement. You telling me that it is my problem and I am at fault don't solve the situation: I don't have time to work on it.

We began the process of deprecating this mode almost a year ago, which we specifically brought to your attention [in this PR](https://github.com/lightningnetwork/lnd/pull/1843). It was laid out very clearly that support for this testing feature will be removed, and since then _no progress_ has been made on migrating existing users away from it.

At the same time, I do acknowledge that migrating existing users is more involved, but as far as I can tell there hasn't even been an attempt to ensure the safety of _new_ users being onboarded to BTCPay. Instead, the number of potentially affected users continues to grow a year later, rather than quarantining the issue to early adopters.

> Note that our change will make the official LND incompatible with BTCPay providing the seed, so people will NEED to run our fork for having the word seed in BTCPay UI. (they already do if they use the docker deploy)

We planned a long deprecation cycle instead of removing it immediately _specifically to support this project_ and offer time to come up with a solution, fully aware of the limited resources available. If the status quo continues and nothing is done, BTCPay will have to maintain `noseedbackup` in their fork either way when it is ultimately removed.

TBH the fact that no progress has been made thus far makes me question whether we should continue to postpone the removal of `noseedbackup` on BTCPay's behalf.

**Comment by NicolasDorier** [comment]
> Why would it refuse to start? You can detect which version of lnd it is (based on the deployment) and conditionally apply the flag.

@Roasbeef all deployments run same version of LND. If we update LND, this new version of LND is also updated, there is no way to update some part of the stack and not other without architectural change which take months.

>  but I would consider the safety of user funds to be a very high priority for any Bitcoin project that involves a wallet or keys

Speaking of which would be nice then that you fix or remove the buggy cookie file authentication in LND. Right now if Bitcoin core reboot while LND use it and that LND use cookiefile authentication, LND does not crash but stop monitoring the chain, leaving open to steal. (though out of topic)

>  some "SFYL" messages ready,

Yes I already do basically telling them they will lose their money and they need to be tech savvy to recover or see the documentation of the implementation they are using.

But even worse than risking a loss, is updating and getting locked out of their fund.

> TBH the fact that no progress has been made thus far makes me question whether we should continue to postpone the removal of noseedbackup on BTCPay's behalf.

You can remove it if we are the only problem. We already maintain or own fork of LND for other reasons. (see my PRs that did not get merged)

My plan on https://github.com/btcpayserver/btcpayserver-docker/issues/179#issuecomment-518559902 will allow to have seed for new deployment, while not breaking current user when they update.

**Comment by Roasbeef** [comment]
You can even just do a new update that includes the seed, then tell old users to use RTL to close all their channels, and send those funds to a new wallet. Then create a new instance.

**Comment by NicolasDorier** [comment]
But the old users will update BEFORE seeing my warning. That's the problem, when they updated this is too late, their LND will not start.

**Comment by Roasbeef** [comment]
> Yes I already do basically telling them they will lose their money and they need to be tech savvy to recover or see the documentation of the implementation they are using.

As far as I've seen, this is noticeably _absent_  from all public promotion of BTCPay.

**Comment by Roasbeef** [comment]
> That's the problem, when they updated this is too late, their LND will not start.

Yep, then they'll complain and realize they need go back to the prior version and sweep their funds. There's no way around this, you _need_ to communicate to users the current state of the safety of their funds.

**Comment by NicolasDorier** [comment]
![image](https://user-images.githubusercontent.com/3020646/62674692-79de4980-b9de-11e9-8ed9-df32d62ccfca.png)

This is on the config page

**Comment by NicolasDorier** [comment]
> Yep, then they'll complain and realize they need go back to the prior version and sweep their funds. 

1. there is no way to downgrade without command line
2. I don't want them flooding the chat in panic mode all at once
3. It would require that I make sure that things can be downgradable, if BTCPay make DB changes this is not possible

**Comment by woutersamaey** [comment]
Some ideas that may help:
- What if we release BTCPay as a new version 2.0 and inform users about it in this light? 1.x users will remain in the same boat. 
- If communication beforehand is an issue, we can simply release a small update now that has a simple notification system (I.e. curl a URL from the docs and if not empty, show at the top of the Admin UI). People will need to do the small upgrade to see the message, sure, but it is helpful and low effort. This also buys you time.
- Personally I do think users have been warned more than enough and as this is experimental software. Sometimes experiments need to be restarted. It should not upset anyone if we close all their channels and move their funds to safety, but they need to re-add channels after.
- Do we really need a full UI for this? Can’t we inform the user (and ask him some Y/N questions during the Docker deployment  update script? Or are you also considering non-dockers to be an issue?

**Comment by Kukks** [comment]
I don't think this should warrant a major release on its own and forcing (
instead of allowing to do it on their own at the very least if we cannot
find a graceful upgrade path) to close all their channels is definitely not
acceptable. I would be furious if a simple version bump wiped out all my
channels( and potentially funds) just because of this issue.

**Comment by woutersamaey** [comment]
@Kukks Forcing is definitely not what we want. That’s why I suggested a different version scheme, allowing people to choose to stay or upgrade. The choice can be presented on the command line  (for docker deployments), but I guess non-dockers are much harder. These were just some ideas. Will shut up now.

**Comment by Kukks** [comment]
> @Kukks Forcing is definitely not what we want. That’s why I suggested a different version scheme, allowing people to choose to stay or upgrade. The choice can be presented on the command line (for docker deployments), but I guess non-dockers are much harder. These were just some ideas. Will shut up now.
@woutersamaey you don't need to shut up haha. We're all neutral and equal in this discussion :)

**Comment by maltokyo** [comment]
Looks like I’d better start closing my channels and sweeping 😅. It was fun while it lasted and I learned a lot so far 👍🏻 I will implement again when I have certainty that I can really use with larger amounts and can back them up. 

I really wish I could help, and while I’ve grown to consider myself a “power user” I can’t contribute in the coding part. Maybe I can help to document this upgrade?

**Comment by woutersamaey** [comment]
I may be able to help with the backup and rollback to a previous/specific version for docker installs. Looks like this will come in handy anyway.

**Comment by NicolasDorier** [comment]
@woutersamaey no need of complicated solution look my solution here:

https://github.com/btcpayserver/btcpayserver-docker/issues/179#issuecomment-518680623

This is like half a day of work without any work on btcpay side aside of reading a file.

**Comment by NicolasDorier** [comment]
> there hasn't even been an attempt to ensure the safety of new users being onboarded to BTCPay. 

But HOW, let me know HOW. I submitted a solution https://github.com/btcpayserver/btcpayserver-docker/issues/179#issuecomment-518680623 after @maltokyo made me think we can make it "opt-in" for old user and default for new users. It requires we do it on our own fork, but that's ok.

**Comment by Roasbeef** [comment]
> I don't want them flooding the chat in panic mode all at once


This seems to be the main pain point on y'alls end. We can help slightly with the support, but I think it's unavoidable as it's a rather major change.

**Comment by Roasbeef** [comment]
> There is no way to differentiate "old user" from "new user", people update all to same version at their own leisure

Is there no versioning information stored in BTCPay's database? As far as distinguishing, you can try to start `lnd` w/ and w/o the flag and handle that there, though this isn't very elegant. 

> I don't care how you think I must use LND, I have a problem and you don't help me so until you help me to get this shit fix, this will continue to drag both of our time.

This all arose due to BTCPay using a flag that we discouraged from the start (`noseedbackup). We've offered help in the form of an update path, and have given information on how to do it safely. 

> It makes BTCPay incompatible with the official LND, but there is nothing we can do more at that point. We have limited resources, we are not a LND wallet,

I don't think this is a good decision, it _removes_ safety features (SCB), and will put your fork further behind mainline `lnd`, which will make maintaining it more difficult for y'all.

**Comment by Roasbeef** [comment]
>  And I told you exactly what was the problem, and this would be easy for you to figure solution out.

If this is about the TLS feature, we opted to have a safe default for all/most users. AFAIK, y'all use an nginx proxy to handle TLS. This is till compatible with lnd's default TLS RPC, as the nginx proxy just need to be given lnd's cert, then it can proxy as normal, offering a Let's Encrypt cert to the regular endpoint.

**Comment by Roasbeef** [comment]
> Assume same behavior as "noseedbackup=1"

Can you elaborate on this?

**Comment by Roasbeef** [comment]
We've planned to do #1 on in your. As for the rest: 
> When a new wallet is created, save the seed mnemonic next to the wallet.db
> Expose the seed mnemonic file through BTCPay.

Why do this require making a fork of `lnd`? If you programmatically instantiate `lnd` [using this call](https://api.lightning.community/#genseed), and [this call](https://api.lightning.community/#initwallet). Then there's no need to make a new fork or read from disk, as you can script this using bash, C# (gRPC), or over the REST interace. The first call gives you a seed, and the second one commits it to `lnd`. As a result there's no UI needed and the user is able to obtain the seed later. You may also want to generate a strong passphrase for the user's wallet unlock as well. 

I understand that you went with the current model because it was easier to deploy, and also easier for the user as they didn't need any knowledge of what a seed is. However, this is the classic security vs usability tradeoff, and in this case, I think the tradeoff went in the wrong direction. Easy buttons are great, but they require you to trade of UX for security.

**Comment by NicolasDorier** [comment]
> Can you elaborate on this?

@Roasbeef so the idea is that we will do a mode which:

1. Does not crash existing noseedbackup users
2. For new users, it will save the seed inside a file next to the wallet db with the default password
3. We will have an "auto unlock" so we don't have to create external tooling to unlock automatically the lnd wallet
4. In BTCPay, just read the seed file and present it in the UX

**Comment by Roasbeef** [comment]
Rather than continue to manage your own fork, wouldn't you rather follow the path I outlined above? You'd also be able to code C# (or w/e) rather than Go as well. 

As for old users, you can resolve this on the docker level:
  * Start using the w/o flag 
  * Do a health check (hit get info, it'll fail all together as no RPC is up if they need the flag)
  * If the health check is succesful, then no action (you can also skip this if you have versioning info available)
  *  If it failed, then start w/ the old flag

**Comment by NicolasDorier** [comment]
@Roasbeef we are already maintaining our fork for other PR that you don't want to merge.

Making a health check and coding a WalletUnlocker seems more painful to us than changing the Go code. If I am correct, our change will require less then 50 lines of Go code, so that is easily maintainable.

Also, I can't use gRPC in C# as long as C# gRPC lib rely on a native binary which will break on different arch. (This should change in .NET Core 3.0 but we are far from migrating, also I don't really like introducing one dependency for only a single call)

**Comment by maltokyo** [comment]
Ok. So I closed all 13 channels and swept all funds out. Ready to get any updates etc that may need testing. I won’t begin again until funds are backed up properly.

**Comment by Roasbeef** [comment]
@NicolasDorier you can use REST, no need for gRPC. Also if you're starting nodes /w `--noseedbackup`, then you'll _need_ to interact with the `WalletUnlocker` in order to initialize the node. Otherwise, it'll just sit there on start up, waiting for a new seed.

**Comment by NicolasDorier** [comment]
Cool, I was not seeing the wallet unlocker on https://api.lightning.community/rest/

> Otherwise, it'll just sit there on start up, waiting for a new seed.

Yes, except if I modify this part, should not be huge deal to maintain.

**Comment by maltokyo** [comment]
@NicolasDorier anything I need to test?

**Comment by NicolasDorier** [comment]
@maltokyo not to test but to code as well. :p

**Comment by maltokyo** [comment]
I would if I could ;(

**Comment by NicolasDorier** [comment]
You definitely can... with enough time :p

**Comment by maltokyo** [comment]
Hehe. Right now I’d cause you more pain than I can help. But if there are other ways to help let me know!

**Comment by maltokyo** [comment]
@NicolasDorier I have just setup a new server. And temporarily changed the lnd docker compose fragment to noseedbackup=0 directly in the file before first setup. Now node is syncing. Will this be enough to get the seed later? Next step is to work out how I actually will get it.

**Comment by maltokyo** [comment]
So, the workaround worked well, I got the LND wallet generation dialogue and could save my LND CIPHER SEED.

All that is needed is, on a fresh install (before running for the first time) change the `noseedbackup=1` to `noseedbackup=0` in file:
`btcpayserver-docker/docker-compose-generator/docker-fragments/bitcoin-lnd.yml`

Then after btc chain finishes syncing run command: `./bitcoin-lncli.sh create`

I guess after you do this once, it doesn't matter after that if upgrades change it back to `noseedbackup=1` in the yml file above.  As this is only relevant if you are creating a wallet anyway.   Also you will need to enter a passphrase for LND wallet each time you start, but that is not a big deal.  All in all, the workaround is 5min extra work for a new install.

I think you should make all new users do this, even the "hard way" for now.  This would make your unbacked-up userbase stop expanding at least.

**Comment by woutersamaey** [comment]
@maltokyo, what are the steps if you have a previous setup, but never used LND on that install, please?

**Comment by maltokyo** [comment]
Thats even easier.  

1. change the `noseedbackup=1` to `noseedbackup=0` in file:
`btcpayserver-docker/docker-compose-generator/docker-fragments/bitcoin-lnd.yml`

2. then enable LND with these two commands:
```
export BTCPAYGEN_LIGHTNING="lnd"
. ./btcpay-setup.sh -i
```

3. Then create wallet with `./bitcoin-lncli.sh create`

**Comment by woutersamaey** [comment]
@maltokyo Actually I meant that I already had done step 2, but have not enabled lightning in the UI, nor put or received any funds or created channels. So in other words, I’m ok if LND is reset or lost.

**Comment by maltokyo** [comment]
in that case, first srop the server, and then obliterate the files in /var/lib/docker/volumes/generated_lnd_bitcoin_datadir and /var/lib/docker/volumes/generated_lnd_bitcoin_rtl_datadir

and rerun entire `export BTCPAYGEN_LIGHTNING="lnd"` and `. ./btcpay-setup.sh -i`

I have not tested it, but it may work.

**Comment by NicolasDorier** [comment]
@maltokyo you should not change the fragment. Instead, you should do your own fragment (see the documentation in btcpayserver-docker) as you will be unable to update otherwise.

I can't just change from 0 to 1 for two reasons:
1. It would require user with 0 to do some SSH hackery to get back their money
2. It would require work for auto unlocking, so that if the server restart, you don't have to manually unlock.

If you can solve those two problems then I can switch by default.

**Comment by maltokyo** [comment]
Hi @NicolasDorier 
Thanks for getting back.  I understood its a workaround, and the limitations.  I can upgrade, as even if an upgrade now switches noseedbackup to "1", it will not affect me, as I have already created the wallet with a  cypher seed.

What I mean is, that you should be able to detect a brand new install on the server (where there is no wallet in existence) and simply then tell the user (in your setup instructions here on github) to open the command line and run a couple of commands.

**Comment by NicolasDorier** [comment]
> simply then tell the user (in your setup instructions here on github) to open the command line and run a couple of commands.

1. I will never lock the user out of their funds after an update
2. I will never requires users to run command lines if things are not broken
3. Detecting a brand new install is not easy thing, PR welcome if you have time to find a way.

**Comment by maltokyo** [comment]
I get all you’re saying. I’m simply letting people know that the workaround
worked. I’ll stop giving further suggestions and ideas for change, as it
feels unappreciated / unwanted somehow. I’m sure those of you who are good
at coding will work it out one day. Coding part is not possible for me at
the moment, as mentioned above.

**Comment by woutersamaey** [comment]
@NicolasDorier what if we just ask users if it is okay to reset their LND (including a thorough explanation of why and the consequences)?
If we can’t detect the need in software, we can ask the user. I guess there will be plenty of installs that have not used LND yet and would not mind a reset. Others could be given the opportunity to switch to the new setup.

What would the auto-unlocking look like in your experience? Please give sufficient detail so I can follow and maybe think about (help) building that.

**Comment by NicolasDorier** [comment]
@woutersamaey I can't break it down in steps (I thought just removing the locking feature by changing LND in our fork is the simplest).

> what if we just ask users if it is okay to reset their LND (including a thorough explanation of why and the consequences)?

Well if you find a way to do that that don't lock them out of their instance when they update, why not. But I can't tell you how.

**Comment by NicolasDorier** [comment]
Our fork is here: https://github.com/btcpayserver/lnd/tree/dockerfile2 so feel free to try to do it

**Comment by NicolasDorier** [comment]
> I’ll stop giving further suggestions and ideas for change, as it
feels unappreciated / unwanted somehow.

@maltokyo , this is not the case

**Comment by junderw** [comment]
Re: detecting fresh install / version

Can't you just add a version file in the docker volume?

Then when someone updates / installs:

1. If no version file the version is < whenever you added it.
2. If version file, it's the version written in the file.
3. If no volume (ie. the proper folders are not there) it's a new install.

**Comment by junderw** [comment]
imo LND's backup method (including seed) is superior to the other lightning implementations. (Though Electrum is also looking promising)

It's a shame that new users of BTCPayServer can't utilize it and the number of users stuck in limbo just keeps growing and growing.

I understand Nicolas's side, and I understand Laolu's side. Both are correct, and it looks like everything is in a stalemate.

Laolu adding xprv export to LND will encourage other projects to do hackey things that are not good for the user. (Though I understand why BTCPay thinks this is the only option for existing users without using command line)

BTCPay adding some sort of versioning that prevents update for existing users of LND but letting new users of LND have the seed and use backup feature is difficult to code, requires time, and is hard to work about with the way BTCPay is currently setup.

This whole situation saddens me quite a bit.

Anything I can do to help fix this issue and allow users going forward to use LND with its full capabilities (aezeed awesomeness and channel backups) I am all ears and willing to help.

**Comment by NicolasDorier** [comment]
@junderw there is a solution we can add to our fork https://github.com/btcpayserver/btcpayserver-docker/issues/179#issuecomment-520198428 , if you have time to talk and look into it, we can do.

**Comment by junderw** [comment]
@NicolasDorier What does "create a new mode" mean?

Like "run this command to install BTCPay+clightning, this command to install BTCPay+LND_OLD, and this command to install BTCPay+LND_NEW"?

IMO Roasbeefs method of tapping the getinfo HTTP REST API and if it fails, restart LND with the noseedbackup flag... seems much easier... but I would be up for either one.

As long as new users can get their seeds, we can just encourage non-techie users to just close down all channels manually via RTL and restart their BTCPay instance from scratch, get their seed and have automated backups.

There is a huge incentive to do this, so unless they are a combination of lazy / not-doing-enough-business-to-merit-backups-anyways-in-their-mind they will probably switch.

**Comment by NicolasDorier** [comment]
> Like "run this command to install BTCPay+clightning, this command to install BTCPay+LND_OLD, and this command to install BTCPay+LND_NEW"?

I don't want to do this. It involves too much support stress and way to lose money.
I can only warn users that does not have the seed in the LND page of BTCPay that they should close channels and move funds, and run a specific command to delete their LND instance. (pointing to a doc which explain step by step)

I don't want to support two modes. New users should have the seed, old user does not. There should have no way to run into "old user mode" via a separate command.

> IMO Roasbeefs method of tapping the getinfo HTTP REST API and if it fails, restart LND with the noseedbackup flag... seems much easier... but I would be up for either one.

Maybe. Though you need to do it in bash, through the `docker-entrypoint.sh`. My guess is that it is easier to code in Go in our fork.

> There is a huge incentive to do this, so unless they are a combination of lazy / not-doing-enough-business-to-merit-backups-anyways-in-their-mind they will probably switch.

Sure, I just don't want to force them, if they don't care. Personally I have around 1K in my LND node, I don't have the seed, I don't care about the seed so I would not migrate. If I lose 1K, that's fine by me.

**Comment by junderw** [comment]
> I don't want to do this. It involves too much support stress and way to lose money.
> I can only warn users that does not have the seed in the LND page of BTCPay that they should close channels and move funds, and run a specific command to delete their LND instance. (pointing to a doc which explain step by step)
> 
> I don't want to support two modes. New users should have the seed, old user does not. There should have no way to run into "old user mode" via a separate command.

ok, my original question was "What does "create a new mode" mean?" and I gave an idea of what I thought you meant.

I understand my idea of what you mean was wrong, but could you explain what a "we will do a mode" is referring to? (from your comment that you gave me the link to)

**Comment by NicolasDorier** [comment]
@junderw yes basically the following:

Let's call this new mode `noseedbackup=2`. We would set all users to `noseedbackup=2`.

If starting with `noseedbackup=2`: 

* No wallet is detected => Create the wallet, with a seed, and put the seed in a clear text file (`seed.txt`)next to the wallet.db.
* Wallet is detected and no `seed.txt` => Assume same behavior as `noseedbackup=1`
* Wallet is detected and `seed.txt` is present => Unlock the wallet automatically using `seed.txt` and `hello` password.

**Comment by junderw** [comment]
@NicolasDorier 

- noseedbackup is a bool in the code currently. 2 is not a valid value.
- Maintaining the LND fork with a complete rewrite of the init code (which is what you are planning) is much harder to maintain than maybe 5-10 lines of bash code that will work as long as the init REST interface doesn't change. (@Roasbeef I'm guessing can chime in on the stable-ness of these interfaces)
- Regardless of this feature or other features, adding a version file to the docker volume seems like a better way to manage versioning rather than "if x file exists, this, if y file exists, that"

**Comment by NicolasDorier** [comment]
> noseedbackup is a bool in the code currently. 2 is not a valid value.

This can be changed in our fork. (or just using a different config flag)

> Maintaining the LND fork with a complete rewrite of the init code (which is what you are planning) is much harder to maintain than maybe 5-10 lines of bash code that will work as long as the init REST interface doesn't change.

I am not sure about that, but if you find a better way to do it in the docker entrypoint, try and let me know.

> Regardless of this feature or other features, adding a version file to the docker volume seems like a better way to manage versioning rather than "if x file exists, this, if y file exists, that"

We can't assume such versioning file exists, as right now it does not exists and we don't know from which version users are upgrading from. So even with a versioning file, we need to have a check `if x file exists`.

**Comment by junderw** [comment]
> But in anyway, we will need to change the LND code

I looked at your changes (here's a github link): https://github.com/btcpayserver/lnd/compare/add905d..3f62988?w=1

Ignoring the extra dockerfiles (which really isn't a patch of lnd) you have maybe changed about 10-20 lines and the patch boils down to "disable TLS" and "disable strict RPC user/pass checks" and the changes are isolated into areas not likely to have conflicts moving forward.

To fix the startup methods moving from a two step method to a single step with auto-generating seed.txt you will easily need to change a couple 100s of lines... these lines will be scattered through files that are changed often.

I think using bash logic is much quicker... I think you will realize this once someone actually sits down and tries to change this part of LND, it is a lot more work than adding a config param or adding an if statement... whereas I think bash will be just a couple of if statements and maybe some sleep + while combos in the case of retry.

I think I might have some time next week

**Comment by NicolasDorier** [comment]
Sure, you can try. I am not convinced it is that hard in Go. I don't have to keep the `noseedbackup=0` and `=1` to work at all. If it is too complicated to fallback, I can just crash LND instead for example.

**Comment by rockstardev** [comment]
OK, migration script is almost ready... I'm taking final round of feedback here:
https://github.com/btcpayserver/lnd/pull/4

Once deployed new script will maintain existing LND deployments and user will be warned in UI to migrate... here is preview of how it will look:

![image](https://user-images.githubusercontent.com/5191402/66977786-ddcf5f00-f06c-11e9-96bb-152af252e917.png)

New LND deployments will be auto initialized with seed and then on subsequent starts LND wallet will be auto unlocked so it can be used by BTCPayServer. User will be able to obtain his seed by simply reading plain text file placed next to `wallet.db`. If they want they can then remove the seed as it's not needed for future wallet unlocks.

**Comment by rockstardev** [comment]
Final call for review of that autoinit bash script... we are moving to testing soon. @Roasbeef @cfromknecht @maltokyo any feedback?

**Comment by maltokyo** [comment]
Looks great for me.  I already run my node with noseedbackup=0 from when I set it up by creating a custom lnd.yaml file.  I will have to work out how to migrate back over to standard once this PR is incorporated.
